frontend/src/components/SearchForm.js

```diff
@@ -1,21 +1,30 @@
-import { Fragment, useState } from "react";
+import { Fragment, useState, useEffect } from "react";
 import '../sass/components/searchform.scss';
 import '../sass/base/typography.scss';
 import icon from '../assets/search-icon.png';
 
-const SearchForm = () => {
+const SearchForm = ({generateResultsData}) => {
     const [radioDifficulty, setRadioDifficulty] = useState('beginner');
```

Comment:
Consider using the `useReducer` hook. Find a demo below:

```javascript
import { Fragment, useEffect, useReducer } from "react";
import '../sass/components/searchform.scss';
import '../sass/base/typography.scss';
import icon from '../assets/search-icon.png';

// This is the initial state for the useReducer hook
let initialState = {
    difficulty: ""
};

// This is the reducer function for the useReducer hook
// it takes the previous state and dispatced action as params
const difficultyReducer = (state, action) => {
    //use a switch statment to check the type of action
    //the user dispatched
    switch (action.type) {
        // if the user clicked beginner, a BEGINNER action is dispatched and so on
        case "BEGINNER":
            //depending on the dispatched action, we return the updated state
            return { ...state,  difficulty: 'beginner'};
        case "INTERMEDIATE":
            return { ...state, difficulty: 'intermediate'};
        case "ADVANCED":
            return {...state, difficulty: 'advanced'};
        default:
            return {difficulty: ""};
    }
}


const SearchForm = ({generateResultsData}) => {
    // initialise the useReducer hook which returns the state
    // and a dispatch method
   const [radioDifficulty, radioDifficultyDispatch] = useReducer(difficultyReducer, initialState);

   // we call this function whenever a user clicks any radio button
   // this function receives the event
   const getRadioDifficulty = (e) => {
       // destructure the event.target object
       const { value } = e.target;

       // we check the value of the event target
       if (value === 'beginner') {
            // we dispatch an action to the reducer
            radioDifficultyDispatch({
                // we set the type of action dispatched
                type: "BEGINNER",
                // we add a payload
                payload: value
            });
        } else if (value === 'intermediate') {
            // we dispatch an action to the reducer
            radioDifficultyDispatch({
                // we set the type of action dispatched
                type: "INTERMEDIATE",
                // we add a payload
                payload: value
            });
        } else {
             // we dispatch an action to the reducer
             radioDifficultyDispatch({
                // we set the type of action dispatched
                type: "ADVANCED",
                // we add a payload
                payload: value
            });
        }
   }

    return (
    <Fragment>
         <div className="search-form">
            <div className="text-input-container">
                <label htmlFor="name">Name</label>
                <input 
                    type="text" 
                    id="name" 
                    name="name" 
                    placeholder="Search for an app idea or simply choose the difficulty level below..."
                /> 
                <img className="search-icon" src={icon} alt="search"/>
            </div>

            <div className="radio-bar-container">
                <div className="radio-choice-container">
                    <input 
                        type="radio" 
                        id="beginner" 
                        name="select" 
                        value="beginner" 
                          //check the current state
                        checked={radioDifficulty.difficulty === 'beginner'} 
                        onClick={getRadioDifficulty}
                        readOnly
                    />
                    <label htmlFor="beginner">
                        <h2 className="heading-secondary heading-radio">Beginner</h2>
                        <p>Developers in the early stages of their learning journey. Those who are typically focused on creating user-facing applications.</p>
                    </label>
                </div>
                <div className="radio-choice-container">
                    <input 
                        type="radio" 
                        id="intermediate" 
                        name="select" 
                        value="intermediate" 
                        //check the current state
                        checked={radioDifficulty.difficulty === 'intermediate'} 
                        onClick={getRadioDifficulty}
                        readOnly
                    />
                    <label htmlFor="intermediate">
                    <h2 className="heading-secondary heading-radio">Intermediate</h2>
                        <p>Developers at an intermediate stage of learning and experience. They are comfortable in UI/UX, using development tools, and building apps that use API services.</p>
                    </label>
                </div>
                <div className="radio-choice-container">
                    <input 
                        type="radio" 
                        id="advanced" 
                        name="select" 
                        value="advanced" 
                          //check the current state
                        checked={radioDifficulty.difficulty === 'advanced'} 
                        onClick={getRadioDifficulty}
                        readOnly
                    />
                    <label htmlFor="advanced">
                    <h2 className="heading-secondary heading-radio">Advanced</h2>
                        <p>Developers who have all of the above, and are learning more advanced techniques like implementing backend applications and database services.</p>
                    </label>
                </div>   
            </div>      
         </div>
    </Fragment>
)}

export default SearchForm;
```

Feel  free  to refactor it so that it is DRY.

Reply:
```javascript
import { Fragment, useState, useCallback } from 'react'
import SearchForm from './SearchForm';
import ResultsList from './ResultsList';
import '../sass/components/main.scss';

function Main() {
    //declare the results state
    const [resultsData, setResultsData] = useState([]);

    // declare the subscriber function to pass to child component
    // useCallback hook to cache the function to prevent
    // unnecessary re-render of the searchForm component
   const setResultsDataHandler = useCallback((value) => {
        setResultsData(value);
    }, []);
    
    return (
        <Fragment>
            <main className='main'>
     
            <h1 className='heading-primary'>abb - always be buildings</h1>
                <SearchForm setResult={setResultsDataHandler} />
                <ResultsList resultsData={resultsData} />
            </main>
        </Fragment>
    )
}

export default Main
```

Reply:
```python
import { Fragment } from "react";
import '../sass/components/resultslist.scss';
import '../sass/base/typography.scss';
import Result from './Result';

const ResultsList = ({resultsData}) => (
    <Fragment>
        <div className="results-container">
            {resultsData.map(app =>                   
                    <Result 
                        key={app._id}
                        name={app.name} 
                        tier={app.tier} 
                    />)
            }   
        </div>
    </Fragment>
)

export default ResultsList;

```